components/autofill/src/sync/address/incoming.rs

```diff
@@ -0,0 +1,1337 @@
+    mirror: Option<RecordData>,
+) -> Record {
+    let i = serde_json::to_value(&incoming).unwrap();
+    let l = serde_json::to_value(&local).unwrap();
+    let mut merged_value: RecordData;
+    let mut merged_record = Map::new();
+
+    let field_names = vec![
+        "given-name",
+        "additional-name",
+        "family-name",
+        "organization",
+        "street-address",
+        "address-level3",
+        "address-level2",
+        "address-level1",
+        "postal-code",
+        "country",
+        "tel",
+        "email",
+    ];
+
+    for field_name in field_names {
+        let incoming_field = i.get(field_name).unwrap().to_string();
+        let local_field = l.get(field_name).unwrap().to_string();
+        let is_local_same;
+        let is_incoming_same;
+
+        match &mirror {
+            Some(m) => {
+                let mirror_field = serde_json::to_value(&m)
+                    .unwrap()
+                    .get(field_name)
+                    .unwrap()
+                    .to_string();
+                is_local_same = mirror_field == local_field;
+                is_incoming_same = mirror_field == incoming_field;
+            }
+            None => {
+                is_local_same = true;
+                is_incoming_same = local_field == incoming_field;
+            }
+        };
+
+        let should_use_local = is_incoming_same || local == incoming.clone();
+
+        if is_local_same && !is_incoming_same {
+            merged_record.insert(String::from(field_name), Value::String(incoming_field));
+        } else if should_use_local {
+            merged_record.insert(String::from(field_name), Value::String(local_field));
+        } else {
+            //local and remote are different
+            return get_forked_action(Record {
+                guid: SyncGuid::new(&guid),
+                data: local,
+            });
+        }
+    }
+
+    merged_value = serde_json::from_str(Value::Object(merged_record).as_str().unwrap()).unwrap();
+    merged_value.time_created = incoming.time_created;
+    merged_value.time_last_used = incoming.time_last_used;
+    merged_value.time_last_modified = incoming.time_last_modified;
+    merged_value.times_used = incoming.times_used;
+
+    Record {
+        guid: SyncGuid::new(&guid),
+        data: merged_value.clone(),
+    }
+}
+
+fn get_forked_action(local_record: Record) -> Record {
+    let mut local_record_data = local_record.data;
+    local_record_data.time_created = Some(Timestamp::now());
+    local_record_data.time_last_used = Some(Timestamp::now());
+    local_record_data.time_last_modified = Some(Timestamp::now());
+    local_record_data.times_used = Some(0);
+    local_record_data.sync_change_counter = Some(1);
+
+    Record {
+        guid: SyncGuid::random(),
+        data: local_record_data,
+    }
+}
+
+fn insert_changes(conn: &Connection, changes: AddressChanges) -> Result<()> {
+    let sql = "INSERT OR IGNORE INTO addresses_data (
```

Comment:
This should be `addresses_sync_applied` not `addresses_data`, will be pushing that in the next commit.